### DISCUSSION TITLE
Gnosis Chain for L2 payouts

### DISCUSSION THREAD

**Discussion by rndquu** [ubiquity/ubiquibot#276] [Research]
Ethereum gas prices are high again and withdrawing payouts becomes more and more expensive.

Some folks are already complaining:
<img width="562" alt="Screenshot 2023-04-28 at 18 01 25" src="https://user-images.githubusercontent.com/119500907/235183318-dd1d58b6-72de-4d76-a55d-a44cfa262573.png">

There are many options for L2 solutions (here the term "L2" also includes sidechains). I think we can select any L2 solution which meets the following criteria:
1. Widely adopted solution
2. Low gas prices
3. Fast bridging `Eth mainnet <=> L2` 

Gnosis Chain is not widely adopted but meets other criteria.

### How fast is the bridge

Bridge website: https://bridge.gnosischain.com

It takes 20 block confirmations on ethereum mainnet (~4 minutes) to transfer DAI from ethereum mainnet to gnosis chain. Technically ethereum mainnet's DAI are converted to XDAI on gnosis chain (XDAI is used for gas payments on gnosis chain, 1 XDAI = 1 DAI).

It takes 20 block confirmations on gnosis chain (~2 minutes) to transfer XDAI from gnosis chain to DAI on ethereum mainnet. It requires 2 transactions: 1st transfer transaction is on gnosis chain and 2nd [claim transaction](https://etherscan.io/tx/0x406103f63104f248ba874cadcd4d1d0403205a3edc5c9e1ef382006bdf8cf718) is on ethereum chain (which is pretty expensive right now) 

### Codebase

Right now we support payouts in ERC20 DAI on ethereum mainnet. XDAI in gnosis chain is not a plain ERC20 token, it is used for gas payments (the same way as ETH is used in ethereum chain). While it seems to be natural to support XDAI (which is pegged to DAI) for payouts in gnosis chain it also requires us to maintain 2 codebases for permit2 and [gnosis safe payouts](https://github.com/ubiquity/bounty-bot/issues/274). By "2 codebases" I mean supporting payouts in ERC20 and native gas tokens which adds complexity. That is why it is better to use WXDAI in gnosis chain which is a plain ERC20 token. 

In short `ERC20 DAI in ethereum = ERC20 WXDAI in gnosis chain`. So bounty hunter payouts are going to be in `WXDAI` in gnosis chain and in `DAI` on ethereum mainnet.

### Bounty hunter flow

From a bounty hunter's view everything is the same. He posts `/wallet 0x01` and solves an issue. Payouts (permit2 URLs or automatic safe transactions) are performed in a chain that would be set in partner's config.

### Partner flow

Partner would need to add a `CHAIN_ID` parameter to the bot's config. So partner will be able to use either ethereum mainnet either gnosis chain.

Also notice that while it is technically possible to create 2 gnosis safes with the same address (on mainnet and in gnosis chain) current gnosis safe web UI does not support this feature. From https://help.safe.global/en/articles/5267779-i-sent-assets-to-a-safe-address-on-the-wrong-network-any-chance-to-recover:
```
existing Safe addresses on Eth mainnet re-created on L2/sidechains (and vice versa) do not load up within our UI
```

### How to migrate on the bot's side

Basically we need to:
1. Read `CHAIN_ID` from a partner's config
2. Make permit2 to work with multiple chains
3. Make [gnosis safe transactions](https://github.com/ubiquity/bounty-bot/issues/274) to work with multiple chains

**Comment by 0x4007** [comment]
I think this looks bounty ready. I agree with the `wxDAI` conclusion. 

Did you post it under discussions because you wanted to work through something more before turning it into an official bounty?

**Reply by rndquu** [reply]
> I think this looks bounty ready. I agree with the `wxDAI` conclusion.
> 
> Did you post it under discussions because you wanted to work through something more before turning it into an official bounty?

I actually wanted an opinion from the @ubiquity/development team, last time there was a holy war discussing what L2 solution to use.

I don't know if it is important from the business side but 1 more point to Gnosis Chain integration is that we use gnosis products (Safe for payouts + gnosis chain support) more and more which makes the project a good candidate for Gnosis VC Investment Fund.

**Reply by 0x4007** [reply]
> I actually wanted an opinion from the https://github.com/orgs/ubiquity/teams/development team, last time there was a holy war discussing what L2 solution to use.

From the business side, the only two viable candidates are Optimism (we are in the process of submitting the grant, the next opportunity to submit is "10 May" according to @Hodlatoor) and Gnosis (we are doing research to figure out how to partner with them.) The reason why these are the two strategic partners chosen over the rest is because of their alignment with the Ethereum vision and ethos. Given that Ubiquity DAO is built on the same values, I believe that they will be excellent long term partners.
